### PR TITLE
Global Styles: Fix Separator block color use

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related:
- https://github.com/WordPress/gutenberg/issues/57297

## What?

This PR removes problematic fudging of Separator block style declarations in favour of filtering the actual theme.json data at both the theme and user origin levels within the theme.json resolver. This approach improves color consistency and flexibility across both origins while also making the Separator block styles more filterable and standardized.

## Why?

Global Styles currently only sets a single property in the user origin theme.json data, `background`. This means if a theme sets the `border.color` or `color.text` property, the `background` color isn't applied consistently.

## How?

- Remove hack tweaking style declarations in favour of theme.json data based approach
- Tweak Separator block style data based on data origin (theme vs user) to provide expected behaviour e.g. user color selection in Global Styles being honoured.

<details>
<summary><strong>Detailed Breakdown</strong></summary>

`lib/class-wp-theme-json-gutenberg.php`

- Removed bespoke function fudging style declarations for separator block.
- This is now handled by filtering the theme origin data in the theme.json resolver. See `WP_Theme_JSON_Resolver::get_theme_data()`

`lib/class-wp-theme-json-resolver-gutenberg.php`

- Separator data is now tweaked at both the user (`get_user_data`) and theme (`get_theme_data`) level.
- User origin data is tweak such that if a user has selected a color that color is used in all color paths the Separator block might require (i.e. `color.background`, `color.text`, or `border.color`)
- Theme origin data is tweaked so that if a `color.background` value has been defined it is used as a fallback for `color.text` and `border.color` if they haven't been defined.
- These tweaks to the theme.json data have been done within the resolver here to avoid the overhead of filters.

`packages/block-editor/src/components/global-styles/use-global-styles-output.js`

- The old hack modifying theme.json data on the fly when generating Global Styles output has been removed as it only related to the theme origin data (hence the global styles bug) and is now covered by the theme.json resolver update at the theme origin level.
- An added benefit of tweaking the Separator style data in the resolver is that it becomes filterable too by extenders which is likely expected.

`packages/editor/src/components/global-styles-provider/index.js`

- The Global Styles provider was updated to handle tweaking user origin data in the same fashion as the theme.json resolver.
- This JS change is required because the editor needs to handle unsaved changes to Global Styles that the user has made. This changes aren't passed through the PHP theme.json resolver until they are saved.

`phpunit/class-wp-theme-json-resolver-test.php`

- New unit tests were added to cover the tweaking of Separator style data in the theme.json resolver.

`phpunit/class-wp-theme-json-test.php`

- The old unit test for the bespoke function handling the fudging of separator style declarations in the theme.json class has been removed. This is now covered by the resolver tests.
</details>

<details>
<summary>Initial context from early explorations into a fix</summary>

### Approach
The initial idea was to enforce the user selected color from Global Styles, i.e. `background` in the user origin.  That evolved to perhaps do so in a way that wouldn't mean that it was also saved to the CPT. This didn't really seem achievable as it would mean having to instantiate additional `WP_Theme_JSON_Gutenberg` instances which come with not insignificant overhead.

To simply have something for testing purposes, I tweaked the retrieval of user origin data in the theme.json resolver class and where the user and theme origin data is merged on the JS side. If we pursue an approach of enforcing the background color across multiple properties, the points at which it is done in JS and PHP should be the same e.g. during merge, or retrieval of user data etc.

The same consistency should also be applied to the existing fudging of separator data in `useGlobalStylesOutput` and the style declarations in `WP_Theme_JSON_Gutenberg`.

### Alternatives

1. On the PHP side, the Separator block could add a `wp_theme_json_data_user` filter and set things there. The equivalent in JS could be done in `useGlobalStylesContext` or `useGlobalStylesUserConfig`.
2. Another hack could be to add Separator specific code to the Global Styles color panel. When setting the background color there, update all the appropriate style values. Then there'd be no tweaking before merging required.

</details>

## Testing Instructions

1. Activate a theme that defines Separator styles using border or text color e.g. TT4 or TT5.
2. Edit a post, page, or template, adding multiple Separator blocks
3. Assign different block styles to each e.g. dots, wide etc.
4. Navigate to Global Styles > Blocks and set a background color for the Separator block
5. Inspect the Separator blocks in the editor and ensurje that they get all three color properties styled i.e. `background`, `border-color`, and `color`.
6. Save and view on the frontend. 
7. Repeat the process of checking the applied styles. All three should be present there as well.
8. Back in the editor double-check that setting block instance styles on the Separator blocks works as per trunk.

### Replicating original issue on trunk

- Use TT5
- Assign a light or bright color to the Separator block in Global Styles
- Note that you can still see the theme set dark color over the top of the color you selected and not at all for a block with the Dots block style selected

## Screenshots or screencast 

Excuse the bright yellow color in these screenshots, it was the easiest option to highlight the pre-existing theme.json color impacting the Global Styles selection.

#### Before
Screenshot 2024-11-25 at 6 47 36 pm

#### After
Screenshot 2024-11-25 at 6 48 31 pm

### CODE REVIEW

File: lib/class-wp-theme-json-resolver-gutenberg.php [hunk 1]
```diff
@@ -563,6 +563,21 @@ public static function get_user_data() {
...
+		/*
+		 * The Separator block uses different CSS properties for its color depending
+		 * on how it is being rendered e.g. as "content" for the Dots style, or
+		 * as a border etc.
+		 * 
+		 * Uses are only presented with a single color control for background. Any
+		 * selection of a background color should be applied to the other paths
+		 * so it can be honoured.
+		 */
+		$separator_color =  $config['styles']['blocks']['core/separator']['color']['background'] ?? null;
```
Comment (ramonjd) [inline lib/class-wp-theme-json-resolver-gutenberg.php]: This reminds me to brush off the [theme json style-related methods abstraction issue](https://github.com/WordPress/gutenberg/issues/65728)

File: lib/class-wp-theme-json-resolver-gutenberg.php [hunk 2]
```diff
@@ -563,6 +563,21 @@ public static function get_user_data() {
...
+		if ( $separator_color ) {
+			_wp_array_set( $config, array( 'styles', 'blocks', 'core/separator', 'color', 'text' ), $separator_color );
+			_wp_array_set( $config, array( 'styles', 'blocks', 'core/separator', 'border', 'color' ), $separator_color );
+		}
```
Comment (ramonjd) [inline lib/class-wp-theme-json-resolver-gutenberg.php:613]: I like the idea of setting things in the user CPT. 

So, if a taking the example of TT4 and TT5 mentioned over in https://github.com/WordPress/gutenberg/issues/57297#issuecomment-2467042865, or any theme that specifies a border only and not a background, any updates to global styles background colors would overwrite it. 

In principle it seems more robust to do this operation at the source data (the merged object), so that any operation consuming this data wouldn't have to compute the styles all the time.

Speaking hypothetically, what would happen to the use of `update_separator_declarations` in `get_styles_for_block`?

I get the feeling this block will have some funky backwards compatibility issues later in life.

For example, throw in the hypothetical that the separator block gets padding support one day. The border color could different to the background color on purpose. 

I guess either way the existing `update_separator_declarations` would have to be updated to handle such a situation.

All in all, letting themes make their own mistakes, but forcing things at the global styles level appears to be a reasonable compromise.

Thanks for getting this discussion going @aaronrobertshaw 

Hopefully we get some more context from https://github.com/WordPress/gutenberg/pull/44943#issuecomment-2497000560

Comment (aaronrobertshaw) [reply lib/class-wp-theme-json-resolver-gutenberg.php:613]: > In principle it seems more robust to do this operation at the source data (the merged object), so that any operation consuming this data wouldn't have to compute the styles all the time.

I'm not sure I follow this comment 100%, sorry. 

I have a suspicion this is about filtering only in `get_merged_data`? I looked at that so that the copying of the background value into all the different paths wasn't saved within the Global Styles CPT but this approach would require instantiating additional `WP_Theme_JSON_Data_Gutenberg` instances which in turn instantiate more `WP_Theme_JSON_Gutenberg` instances causing a performance hit I wanted to avoid. This is why it was done after retrieving and decoding the CPT and before applying filters on the PHP side.

This PR still only filters the merged data on the JS side though as I was experimenting with still trying to avoid saving all the different path values and keep it to only the `background` color. Part of the next steps here is to consolidate everything including filtering at the same point on both JS and PHP sides. That might require moving the tweaks from `mergeBaseAndUserConfigs` to `useGlobalStylesUserConfig`.



> For example, throw in the hypothetical that the separator block gets padding support one day. The border color could different to the background color on purpose.

True but I suspect a Group block with no content would just be used instead as that's what the separator effectively becomes. I'd say it's a Spacer but that's been discussed on block support adoption PRs as being "layout-only".

> Speaking hypothetically, what would happen to the use of update_separator_declarations in get_styles_for_block?
> ...
> I guess either way the existing update_separator_declarations would have to be updated to handle such a situation.

The pre-existing hack was kept in this PR just to test the hypothesis that forcing the user selection in Global Styles wouldn't break existing block styles. 

My intention now the hypothesis seems to be proven is to consolidate the handling of the separator styles as much as possible. The declarations hack really seems out of place to me. That should have been a filter of the theme.json data all along in my view, which can be done at the theme origin level. 

So the next step for this PR is to get rid of `update_separator_declarations` from the theme.json style generation and tweak the theme origin's theme.json data instead. The existing hack in `updateConfigWithSeparator` and `useGlobalStylesOutputWithConfig` probably gets removed entirely as theme data would be filtered PHP side before it is passed to the editor via the block editor settings.

That's the plan at least. As the onion layers get peeled back, it might all change.

Comment (ramonjd) [reply lib/class-wp-theme-json-resolver-gutenberg.php:613]: > > In principle it seems more robust to do this operation at the source data (the merged object), so that any operation consuming this data wouldn't have to compute the styles all the time.

> 
> I'm not sure I follow this comment 100%, sorry.

It's long-winded way of saying I agree with your approach here 😄 and typed without re-reading, sorry. 

Overriding at the user data, rather than later in styles declarations, means that anyone grabbing the user data, say, from `WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()` will benefit. 

Unless I'm completely missing the point, which could be the case.

> True but I suspect a Group block with no content would just be used instead as that's what the separator effectively becomes

I'm wondering if there's still semantic value in using `<hr />` element. Anyway, it's a moot point at this stage.

Comment (aaronrobertshaw) [reply lib/class-wp-theme-json-resolver-gutenberg.php:613]: Thanks for clarifying 👍 

> I'm wondering if there's still semantic value in using `<hr />` element. Anyway, it's a moot point at this stage

Now I'm actually thinking of this as an `hr` element rather than a Gutenberg block. I don't believe `hr` elements support `padding`. My understanding is they are a ["void element"](https://developer.mozilla.org/en-US/docs/Glossary/Void_element) so don't have a content box to which `padding` can be applied. If I'm correct on that, we can disregard padding support being added to this block.

In a past Gutenberg life, I played with creating a ["divider" block](https://github.com/WordPress/gutenberg/pull/32664) which allowed different SVG based shapes in a stackable fashion. Something like that might strike a balance between the layout-only Spacer block and the decorative Separator but with more control on padding etc (still minus the semantics though 🙁 )

Comment (ramonjd) [reply lib/class-wp-theme-json-resolver-gutenberg.php:613]: > I don't believe hr elements support padding.

Good call! I didn't bother to check. Makes sense to me 👍🏻

**Comment by aaronrobertshaw** [comment]
In https://github.com/WordPress/gutenberg/pull/67269/commits/e71fd97215bbdd3e2d749a45354aaa01a175b158 I've taken a run at removing the fudging of style declarations for the separator block in favour of a theme.json data based approach.

There's a slight change in logic when it comes to merging the theme origin data (i.e. the separator styles when defined in theme.json). 

Previously, the `color.background` value was only used if both `color.text` and `border.color` aren't set. It's a little more robust if we apply the `color.background` value to the other paths if they are undefined. This way, if a theme author set two of the three paths (including `color.background`), any block styles or rendering of the separator block that relied on the missing color value would get the likely expected background color.

The theme origin tweaking of separator color only needs to happen on the PHP class side in `get_theme_data` as the updated data gets passed to the editor via the block editor settings. 

The user origin data tweaks still need to occur on the JS side too. This is due to the editor needing to handle unsaved selections made by the user in the Global Styles UI.

I've removed the old `update_separator_declarations` function and tests. New unit tests have been added to the theme.json resolver tests. I might have gone a little overboard there as they are quite verbose but for the purpose of this PR and discussion, I'll leave them for now and we can cull, clean, or take a different approach later.

@ramonjd, this isn't a high priority but if you get the chance, let me know what you think of the general approach now. If you think it has legs, I'll take this out of draft status and cast the net wider for alternative ideas.

**Comment by ramonjd** [comment]
> I've taken a run at removing the fudging of style declarations for the separator block in favour of a theme.json data based approach.

I like that the code is now in existing internals. I know it's not always easy in practice, but the fewer bespoke private functions we have in theme.json the better. 

> It's a little more robust if we apply the color.background value to the other paths if they are undefined. This way, if a theme author set two of the three paths (including color.background), any block styles or rendering of the separator block that relied on the missing color value would get the likely expected background color.

These are the changes in `get_theme_data()`? 

I think that's sound reasoning. 

From looking at the changes, it's doing its best to honor the theme.json values and safe-guard against unexpected separator colors. 

Overall, I think it's a good direction and you've balanced the competing priorities. Off the top of my head, no backwards compat issues occur to me, but hopefully wider testing will wash that out.

Thank a lot for persisting with this.

What do you think? Should we try to get eyes on it early in 6.8 cycle?

**Comment by ramonjd** [comment]
Forgot to add that I tested the PR too in TT4 and TT5 and it works as expected 😄

**Comment by aaronrobertshaw** [comment]
> These are the changes in get_theme_data()?

That's correct. I should have been clearer on that point.

There was previously no consideration of which origin the data was coming from. Which in part, was the cause for global styles selections not being honoured.

> What do you think? Should we try to get eyes on it early in 6.8 cycle?

💯 Agreed.

I'll give the PR, its description, and test instructions a bit of a polish and flag for review later this afternoon.

Thanks for all the sanity checking around the approach 🙇

**Comment by aaronrobertshaw** [comment]
I've added a more detailed breakdown of all the changes to the PR description and flagged as ready for review. I'll sort out the backport shortly and then ping for a few reviews.

**Comment by aaronrobertshaw** [comment]
Backport is available (https://github.com/WordPress/wordpress-develop/pull/7927) and changelog added here.

File: lib/class-wp-theme-json-resolver-gutenberg.php [hunk 3]
```diff
@@ -563,6 +584,21 @@ public static function get_user_data() {
...
+		/*
+		 * The Separator block uses different CSS properties for its color depending
+		 * on how it is being rendered e.g. as "content" for the Dots style, or
+		 * as a border etc.
+		 *
+		 * Uses are only presented with a single color control for background. Any
+		 * selection of a background color should be applied to the other paths
+		 * so it can be honored.
+		 */
+		$separator_color = $config['styles']['blocks']['core/separator']['color']['background'] ?? null;
```
Comment (ramonjd) [inline lib/class-wp-theme-json-resolver-gutenberg.php:609]: Noticed over here: https://github.com/WordPress/gutenberg/issues/57297#issuecomment-2513233950

What do you think we should do about gradients? If anything at all?

They are applied to the `background` prop, but don't really have any effect. And `linear-gradient` is invalid for `color` and `border-color` anyway as far as I see 🤔

Comment (aaronrobertshaw) [reply lib/class-wp-theme-json-resolver-gutenberg.php:609]: I'll dig into this one tomorrow and see if there's anything we can smooth out.

**Comment by talldan** [comment]
Gave this a test. In TT4, I'm seeing that the block's instance color settings aren't overriding global style color for the 'Default' or 'Wide Line' block styles variations:
Screenshot 2024-12-04 at 10 09 00 am

It seems like the color and background color are being overridden, but not the border color:
Screenshot 2024-12-04 at 10 10 39 am

File: lib/class-wp-theme-json-resolver-gutenberg.php (hunk 3 above)
Comment (aaronrobertshaw) [reply lib/class-wp-theme-json-resolver-gutenberg.php:609]: I've spent some time on an archaeological dig today around the addition of gradient support to the Separator block. 

The TL;DR is that I believe it was simply mistakenly added. It wasn't referred to at all on the [original PR](https://github.com/WordPress/gutenberg/pull/38428/) and doesn't appear to have been tested in any way either. From what I can tell gradients have been broken since their adoption.

---

So...here's the long version of my current understanding 😅 

- Global Styles will generate styles targeting the `background` CSS property
- The Separator block's `hr` element will inherit `currentColor` for its `color` CSS value
- In the editor, the Separator block has a tiny amount of padding added which exposes some `background`
- On the frontend, the Separator has no padding normally so no background is visible
- Even in the editor where there is some background visible it's overlaid by the `color` property effectively masking any gradient applied here
- If the `background` property has a `linear-gradient` value at all, copying that to `border-color` or `color` results in an invalid CSS rule and falls back to `currentColor`
- The Separator block was updated in https://github.com/WordPress/gutenberg/pull/38428 which skipped color block support serialization but it didn't take gradients into account at all
- The approach in [#38428](https://github.com/WordPress/gutenberg/pull/38428) means 
     - gradient classes aren't considered and remain on the block wrapper 
     - any preset `gradient`  isn't retrieved and processed like the `backgroundColor` value
     - when [`customColor` is determined](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/separator/save.js?plain=1#L17) it only takes into account `style.color.background` forgetting about `style.color.gradient` 
- There's one default theme (TwentyTwenty) that appears to leverage `linear-gradient` while it sort of works the `//` characters don't get the gradient and hovering the block in the editor breaks the `//` onto separate lines
- A theme could style the Separator block's `hr` element to not use border or color, give it a height or padding and then apply any background style including gradients. 

At the end of all of that, I'm not sure if this is fixable to a degree where the Separator block could reliably have gradients applied to it. If it was "fixed" it might only be for use cases where the Separator block has had its default styles overridden or has a block style that makes `color` transparent, removes `border`, and gives it a visible height.

I'd like to see gradient support removed from the block. Given its current broken state and that it has been that way since its inception, I don't think there's much of a backwards compatibility argument to be made.  Removing the gradients support at least lowers complexity and maintenance burden moving forward.

What do you think @ramonjd?

**Comment by aaronrobertshaw** [comment]
Thanks for testing @talldan 👍 

>  I'm seeing that the block's instance color settings aren't overriding global style color for the 'Default' or 'Wide Line' block styles variations:

This was another oversight in the original implementation in https://github.com/WordPress/gutenberg/pull/38428/ in addition to mistakenly (in my opinion) adopting gradient support. Further details from some recent digging can be found in https://github.com/WordPress/gutenberg/pull/67269#discussion_r1868675448.

If my understanding is correct, the noted behaviour is consistent with trunk. My focus for this PR was fixing Global Styles. If we're changing the application of background to border-color on the block instance, it will need a deprecation as well. Do you think this is ok as a follow-up?

File: lib/class-wp-theme-json-resolver-gutenberg.php (hunk 3 above)
Comment (ramonjd) [reply lib/class-wp-theme-json-resolver-gutenberg.php:609]: Amazing digging, thank you!

> I'd like to see gradient support removed from the block. Given its current broken state and that it has been that way since its inception, I don't think there's much of a backwards compatibility argument to be made. Removing the gradients support at least lowers complexity and maintenance burden moving forward.

+1

In any case, since it's a long-standing issue, and doesn't affect the intention of this PR it's a "tomorrow" thing. 

I appreciate the time you spent researching. It'll be useful to whoever picks it up.

Comment (aaronrobertshaw) [reply lib/class-wp-theme-json-resolver-gutenberg.php:609]: If we do go with removing support it should only be deleting a single line from the block.json. Despite that, still worth a separate PR to debate and document the decision. I'll spin that up after I work through the possible follow-up for the [issue](https://github.com/WordPress/gutenberg/pull/67269#issuecomment-2516012104) Dan flagged below.

**Comment by talldan** [comment]
> Do you think this is ok as a follow-up?

Yep, sounds good if you think it's best to split it across multiple PRs.

**Comment by aaronrobertshaw** [comment]
Given it's taken me a while to get back to this, I've pushed what I think could be a fix to this PR for the block instance styles not appearing to take effect correctly in https://github.com/WordPress/gutenberg/pull/67269/commits/407464147ac1b49bbf72156db259e4ce305b602c.

| Before | After |
|---|---|
| Screenshot 2024-12-06 at 6 03 35 pm | Screenshot 2024-12-06 at 6 03 29 pm |

I went with a `0-2-0` specificity style in the block library styles (see https://github.com/WordPress/gutenberg/pull/67269/commits/407464147ac1b49bbf72156db259e4ce305b602c) to avoid needing to override the global styles inline, potentially requiring a deprecation. 

Let me know what you think, whether you see any pitfalls to this approach etc 🙏

File: packages/block-library/src/separator/style.scss [hunk 4]
```diff
@@ -8,6 +8,13 @@
 	border-bottom: none;
 }
 
+// 0-2-0 specificity required to enforce block instance
+// color selections over global styles.
+.wp-block-separator.has-background,
+.wp-block-separator.has-text-color {
+	border-color: currentColor;
+}
```
Comment (ramonjd) [inline packages/block-library/src/separator/style.scss:16]: Just so I understand:

- if the separator block has a background color set, the `color` property will have the same value (thanks to the changes in this PR),
- this rule therefore ensures the border color has the same value as `color`, through `currentColor`

> Gave this a test. In TT4, I'm seeing that the block's instance color settings aren't overriding global style color for the 'Default' or 'Wide Line' block styles variations:

The CSS rule above addresses this issue for me in the editor and frontend:

1. In TT4, change the background value of Wide Line variation in global styles
2. At the block level, apply Wide Line variation, and overwrite the background color

Do block style variation global styles need the same treatment? 🤔 

1. In TT4, change the background value of Wide Line variation in global styles
2. At the block level, apply Wide Line variation

The background color is set, but the [theme.json border color](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-content/themes/twentytwentyfour/theme.json#L740) still affects the appearance. Here's what I'm seeing in in the frontend:

```css
:root :where(.wp-block-separator.is-style-wide--2) {
    background-color: #00ff40;
}

:root :where(.wp-block-separator) {
}
:root :where(.wp-block-separator) {
    border-color: currentColor;
    border-width: 0 0 1px 0;
    border-style: solid;
    color: var(--wp--preset--color--contrast);
}
```

Comment (aaronrobertshaw) [reply packages/block-library/src/separator/style.scss:16]: Thanks for the thorough testing @ramonjd.

> Do block style variation global styles need the same treatment? 🤔

I think they need the same resolution that we perform for the generic block type i.e. copy the background color to the other properties as required.

Comment (aaronrobertshaw) [reply packages/block-library/src/separator/style.scss:16]: In https://github.com/WordPress/gutenberg/pull/67269/commits/de096449303fa2cf1b349c1f94e99a0e17f2f520 I've had a quick go at merging the block style variation colors as done for the general block type.

I ran into some caching issues with the PHPUnit tests which has led to me running out of time to test this fully. There's a chance I might need to put this on the backburner soon unless others can pick it up.

**Comment by aaronjorbin** [comment]
This is attached to a ticket slated for 6.9 - https://core.trac.wordpress.org/ticket/62624

@ryanwelcher @t-hamano can y'all triage if this is still a priority and if not, let's punt the core trac ticket to future release.

**Comment by t-hamano** [comment]
Thanks for the ping!

I feel this PR is not essential for 6.9 and can be punted to a future release. I'm sure that the problem this PR is trying to solve is complex and has been around for a long time. I think it's best to proceed carefully and not rush this PR.

**Comment by aaronrobertshaw** [comment]
Thanks for giving this a nudge 👍 

This PR can definitely wait a little longer. I won't be able to devote further time to dusting it off for a few weeks yet.
